### ISSUE TITLE
embed: document how to use embedding from the Go API 

### ISSUE THREAD

**Issue by mvdan** [cue-lang/cue#3613]
As @DavidGamba points out in https://github.com/cue-lang/cue/discussions/3264#discussioncomment-10708504, it's not at all obvious how to use `CUE_EXPERIMENT=embed` from the Go API. He shows a snippet like:

    ctx := cuecontext.New(cuecontext.Interpreter(embed.New())

This is basically what cmd/cue does.

However, note that unlike `CUE_EXPERIMENT=evalv3` or `CUE_EXPERIMENT=toposort`, simply using `os.Setenv("CUE_EXPERIMENT", "embed")` will not make embedding work out of the box; this is because `cue/interpreter/embed` imports `cue/cuecontext` for various reasons, so `cuecontext` cannot import `embed` to enable it following `CUE_EXPERIMENT`.

This gets trickier once `CUE_EXPERIMENT=embed` is on by default; presumably we would want it to work out of the box with `cuecontext.New` just like it would for `cmd/cue`.

It's a different story if we want embedding to be opt-in for the Go API even though it would always be on for the `cmd/cue` CLI. Personally I would find that a bit puzzling, but we haven't explicitly made a choice as part of the proposal design.

My instinct is that we should make `cuecontext.New` enable embedding out of the box just like `cmd/cue` does, but note that this would require untangling the import cycle caused by it. My instinct is that we want to untangle that anyway; having low-level packages like `internal/filetypes` depend on a high-level public package like `cue/cuecontext` is going to cause other trouble down the line.

**Comment by rogpeppe** [comment]
Relevant issue: https://github.com/cue-lang/cue/issues/3280

**Comment by mvdan** [comment]
Indeed, if we could resolve that and avoid loading/evaluating CUE at init time in the filetypes package, then untangling this cycle would be much more feasible :)

**Comment by rogpeppe** [comment]
> Personally I would find that a bit puzzling, but we haven't explicitly made a choice as part of the proposal design.

One reason someone might not wish this: currently evaluating CUE does not allow that CUE to inspect anything in the local file system. If we enabled embedding by default, it would be possible for CUE code to embed files from the local filesystem, which is a potential security concern.

**Comment by DavidGamba** [comment]
Also, currently enabling embed requires to be inside a cue module. Normal Go API calls allow me to read config files without a module, so if a config file needs embed and I don't want to run within a cue module (e.g. a dotrc file config for a tool ~/.tool.cue) then you need an overlay with a virtual module so that embed is happy.

Something like:

```go
	insts := []*build.Instance{}
	overlay := map[string]load.Source{}

	if virtualCueModuleName != "" {
		modPath := filepath.Join(overlayRootDir, "cue.mod/module.cue")
		overlay[modPath] = load.FromString(fmt.Sprintf(`module: "%s"`, virtualCueModuleName))
		Logger.Printf("overlay: %s\n", modPath)
	}

	lc := &load.Config{
		Package:             packageName,
		ModuleRoot:          ".",
		AcceptLegacyModules: true,
		Dir:                 dir,
		Overlay:             overlay,
	}
	ii := load.Instances(packagePaths, lc)

	insts = append(insts, ii...)

	logInstancesFiles("building", insts)
	vv, err := c.BuildInstances(insts)
	if err != nil {
		return fmt.Errorf("failed to build instances: %w", err)
	}
	for _, v := range vv {
		*value = (*value).Unify(v)
	}

```

**Comment by myitcv** [comment]
@DavidGamba in general we want the [CUE loader](https://pkg.go.dev/cuelang.org/go/cue/load) to be responsible for doing everything. That way, all CUE-related tooling can just load CUE and "get it right". Put another way, we want to avoid custom load mechanisms where possible if people intend their CUE to be loadable by standard tooling (e.g. LSP) or their tool to be used by others on their CUE (e.g. some developer tooling). That said, is there a particular reason you want to support embed working outside of a CUE module? The reasons for making that constraint were that it made defining security properties of general CUE easier to understand.

**Comment by DavidGamba** [comment]
@myitcv using CUE loader works when you are working with "proper" CUE configurations in a CUE module.
My use case is for my general purpose tools, where CUE is just used as a config file language in the same way I would use a toml or a ini file.

For example, I can have a tool doing something inside a proper CUE configuration module and it's config can live in the same directory under a hidden file `.mytool-config.cue`.
In such case I don't want the CUE Go API for my tool to read the module info under `cue.mod` but just focus on the list of files or package names I am providing ignoring the rest.

For one such tool I wanted to embed a json file into my config so it would be backwards compatible with another tool we already had and that was when I ran into embed needing a module to run.

The CUE loader does support overlays so it works for my use case but I am fully aware the LSP would probably not work for that config file since it wouldn't know about it.

The only way I envision this working with the generic CUE cli command and not just through the API is if CUE added an annotation or a keyword to indicate that a package is "standalone" and so the files found in the current dir for that package belong to a virtual or standalone module and not whatever a cue.mod directory that could or could not exist contains.
That then adds the complication of supporting imports if you really want to make standalone modules full fledged, but so far my use cases have bundled the entire schema in the Go binary so I haven't needed to use any imports or anything else than that one embed and the stdlib.
